_posts/2019-05-10-why-projects-need-codenames.markdown

```diff
@@ -0,0 +1,47 @@
+
+Code names embrace that **we don't necessarily know the fit or scope of a system when it's first conceived**. To _not_ employ code names would require accurately predicting a system's eventual function. A project named `artsy-admin` starts to grate when we make the architectural decision to split apart management utilities or introduce more fine-grained administrative roles. An `artwork-taxonomy` label loses meaning when that system expands to include less structured tags
+
+**Branding shifts, products pivot, and companies merge.** Project code names introduce a level of indirection between what engineers build and the labels, URLs, or brands that end users experience. Over a long enough period, these external or superficial changes _will_ happen and risk introducing confusion or just subtle misalignment between form and function. As with software design in general, abstractions can be a powerful tool to [separate concerns](https://en.wikipedia.org/wiki/Separation_of_concerns). Ask yourself: what code changes might be required if your company were to spin off your product or be merged into an acquirer?
+
+**Language is important**, and not just for communication but for shaping our thoughts and assumptions. We frequently find ourselves debating which system should serve as the authority for a given domain model or where to implement a new feature, and project names that overlap with these topics (e.g., `search`, `images`, `suggestions`) would predispose us to certain decisions. Code names free us to focus on the architectural and organizational merits instead.
+
+With this in mind, how should you choose a naming scheme?
+
+### Rules for a project naming scheme
+
+* Avoid implicit value judgments like "new," "next," or "modern." We've all witnessed today's hot project become next year's unloved albatross.
+* Choose a code name scheme that isn't directly related to your technology or business. A flower business using flower names is cute, but breaks down when you want to build a feature that _actually_ is about tulips.
+* Everyone should be able to participate, so avoid industry lingo or anything obscure. Funny story: I work at an art start-up but don't know enough artists to name my projects that way. I struggle even more to pronounce the few projects that are.
+* There should be lots of choices. You'll regret choosing to name your projects after "decathlon sports" or "days of the week" when your team inevitably transitions to microservices.
+* Names should be unique within a company and--ideally--beyond the company's scope. You wouldn't want your `marketing-site` repository to conflict with a contributor's `marketing-site` repository from another organization. Github projects, S3 buckets, Heroku application names, and published libraries all benefit from being globally unique.
```

Comment:
@alloy - as a side-note, was thinking that before the post on RRM gets published perhaps we should think of a better / unique name? A lot of its characteristics are lower-level, as a tool to build tools.

Reply:
Oh goody! I thought we had jointly settled on boring names for these, but in that case I’ll give it a think–I love code-names!

Reply:
Yah that lib needs something cool :)